add scenarios to set several vars at once

kdelee opened the pull request:
With named scenarios, we can encapsulate the 4 or more variables we would otherwise have to set.

Users will still be able to override those individual variables if they so desire.

I will make a follow up PR that will add an option in Jenkins to select from a drop down one of these scenarios.

I've also added a debug task to print out what values have been set based on the scenario.

scenarios:
```
control-cluster
hybrid-cluster
control-cluster-hop-to-execution-plane
hybrid-cluster-hop-to-execution-plane
single-hybrid
legacy-cluster-with-isolated-nodes
legacy-cluster
```

```
for scenario in $(cat ~/scenarios); do ansible-playbook -i localhost, -e ansible_connection=local create.yml -e @tower_vars.yml -e @vars.yml --vault-password=~/tower/towerperfvaultpassword -e scenario=$scenario --tags=debug >> debug.txt; done

PLAY [Create IBM Cloud VPC VSI] **************************************************************************************************************************************************************************************************************

TASK [Print some debug information] **********************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": [
        "Requested Topology:",
        "--------------------------------",
        "Topology: control-cluster",
        "Inventory Template: inventory_cluster_4.1.j2",
        "",
        "Control Plane:",
        "--------------------------------",
        "Control Nodes: 3",
        "Hybrid Nodes: 0",
        "Total Control Plane Nodes to deploy: 3",
        "",
        "Execution Plane:",
        "--------------------------------",
        "Hop Nodes: 0",
        "Execution Nodes: 3",
        "Total nodes to deploy for execution plane: 3",
        "Enable execution nodes: True",
        "",
        "",
        "Isolated Nodes:",
        "--------------------------------",
        "Legacy Isolated Nodes requested: 0",
        "Legacy Isolated Nodes to deploy: 0",
        "Enable legacy isolated nodes: False",
        ""
    ]
}

PLAY [Create IBM Cloud VPC VSI for Tower Isolated Nodes] *************************************************************************************************************************************************************************************

PLAY [Create IBM Cloud VPC VSI for Tower Execution Nodes] ************************************************************************************************************************************************************************************

PLAY [Check Ansible connection to new DEMO VSI] **********************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY [Add team ssh keys] *********************************************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY [Install Tower] *************************************************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   


PLAY [Create IBM Cloud VPC VSI] **************************************************************************************************************************************************************************************************************

TASK [Print some debug information] **********************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": [
        "Requested Topology:",
        "--------------------------------",
        "Topology: hybrid-cluster",
        "Inventory Template: inventory_cluster_4.1.j2",
        "",
        "Control Plane:",
        "--------------------------------",
        "Control Nodes: 0",
        "Hybrid Nodes: 3",
        "Total Control Plane Nodes to deploy: 3",
        "",
        "Execution Plane:",
        "--------------------------------",
        "Hop Nodes: 0",
        "Execution Nodes: 3",
        "Total nodes to deploy for execution plane: 3",
        "Enable execution nodes: True",
        "",
        "",
        "Isolated Nodes:",
        "--------------------------------",
        "Legacy Isolated Nodes requested: 0",
        "Legacy Isolated Nodes to deploy: 0",
        "Enable legacy isolated nodes: False",
        ""
    ]
}

PLAY [Create IBM Cloud VPC VSI for Tower Isolated Nodes] *************************************************************************************************************************************************************************************

PLAY [Create IBM Cloud VPC VSI for Tower Execution Nodes] ************************************************************************************************************************************************************************************

PLAY [Check Ansible connection to new DEMO VSI] **********************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY [Add team ssh keys] *********************************************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY [Install Tower] *************************************************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   


PLAY [Create IBM Cloud VPC VSI] **************************************************************************************************************************************************************************************************************

TASK [Print some debug information] **********************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": [
        "Requested Topology:",
        "--------------------------------",
        "Topology: control-cluster-hop-to-execution-plane",
        "Inventory Template: inventory_cluster_4.1.j2",
        "",
        "Control Plane:",
        "--------------------------------",
        "Control Nodes: 3",
        "Hybrid Nodes: 0",
        "Total Control Plane Nodes to deploy: 3",
        "",
        "Execution Plane:",
        "--------------------------------",
        "Hop Nodes: 1",
        "Execution Nodes: 2",
        "Total nodes to deploy for execution plane: 3",
        "Enable execution nodes: True",
        "",
        "",
        "Isolated Nodes:",
        "--------------------------------",
        "Legacy Isolated Nodes requested: 0",
        "Legacy Isolated Nodes to deploy: 0",
        "Enable legacy isolated nodes: False",
        ""
    ]
}

PLAY [Create IBM Cloud VPC VSI for Tower Isolated Nodes] *************************************************************************************************************************************************************************************

PLAY [Create IBM Cloud VPC VSI for Tower Execution Nodes] ************************************************************************************************************************************************************************************

PLAY [Check Ansible connection to new DEMO VSI] **********************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY [Add team ssh keys] *********************************************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY [Install Tower] *************************************************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   


PLAY [Create IBM Cloud VPC VSI] **************************************************************************************************************************************************************************************************************

TASK [Print some debug information] **********************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": [
        "Requested Topology:",
        "--------------------------------",
        "Topology: hybrid-cluster-hop-to-execution-plane",
        "Inventory Template: inventory_cluster_4.1.j2",
        "",
        "Control Plane:",
        "--------------------------------",
        "Control Nodes: 0",
        "Hybrid Nodes: 3",
        "Total Control Plane Nodes to deploy: 3",
        "",
        "Execution Plane:",
        "--------------------------------",
        "Hop Nodes: 1",
        "Execution Nodes: 2",
        "Total nodes to deploy for execution plane: 3",
        "Enable execution nodes: True",
        "",
        "",
        "Isolated Nodes:",
        "--------------------------------",
        "Legacy Isolated Nodes requested: 0",
        "Legacy Isolated Nodes to deploy: 0",
        "Enable legacy isolated nodes: False",
        ""
    ]
}

PLAY [Create IBM Cloud VPC VSI for Tower Isolated Nodes] *************************************************************************************************************************************************************************************

PLAY [Create IBM Cloud VPC VSI for Tower Execution Nodes] ************************************************************************************************************************************************************************************

PLAY [Check Ansible connection to new DEMO VSI] **********************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY [Add team ssh keys] *********************************************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY [Install Tower] *************************************************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   


PLAY [Create IBM Cloud VPC VSI] **************************************************************************************************************************************************************************************************************

TASK [Print some debug information] **********************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": [
        "Requested Topology:",
        "--------------------------------",
        "Topology: single-hybrid",
        "Inventory Template: inventory_cluster_4.1.j2",
        "",
        "Control Plane:",
        "--------------------------------",
        "Control Nodes: 0",
        "Hybrid Nodes: 1",
        "Total Control Plane Nodes to deploy: 1",
        "",
        "Execution Plane:",
        "--------------------------------",
        "Hop Nodes: 0",
        "Execution Nodes: 0",
        "Total nodes to deploy for execution plane: 0",
        "Enable execution nodes: False",
        "",
        "",
        "Isolated Nodes:",
        "--------------------------------",
        "Legacy Isolated Nodes requested: 0",
        "Legacy Isolated Nodes to deploy: 0",
        "Enable legacy isolated nodes: False",
        ""
    ]
}

PLAY [Create IBM Cloud VPC VSI for Tower Isolated Nodes] *************************************************************************************************************************************************************************************

PLAY [Create IBM Cloud VPC VSI for Tower Execution Nodes] ************************************************************************************************************************************************************************************

PLAY [Check Ansible connection to new DEMO VSI] **********************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY [Add team ssh keys] *********************************************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY [Install Tower] *************************************************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   


PLAY [Create IBM Cloud VPC VSI] **************************************************************************************************************************************************************************************************************

TASK [Print some debug information] **********************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": [
        "Requested Topology:",
        "--------------------------------",
        "Topology: legacy-cluster-with-isolated-nodes",
        "Inventory Template: inventory_cluster_4.j2",
        "",
        "Control Plane:",
        "--------------------------------",
        "Control Nodes: 0",
        "Hybrid Nodes: 3",
        "Total Control Plane Nodes to deploy: 3",
        "",
        "Execution Plane:",
        "--------------------------------",
        "Hop Nodes: 0",
        "Execution Nodes: 0",
        "Total nodes to deploy for execution plane: 0",
        "Enable execution nodes: False",
        "",
        "",
        "Isolated Nodes:",
        "--------------------------------",
        "Legacy Isolated Nodes requested: 2",
        "Legacy Isolated Nodes to deploy: 2",
        "Enable legacy isolated nodes: True",
        ""
    ]
}

PLAY [Create IBM Cloud VPC VSI for Tower Isolated Nodes] *************************************************************************************************************************************************************************************

PLAY [Create IBM Cloud VPC VSI for Tower Execution Nodes] ************************************************************************************************************************************************************************************

PLAY [Check Ansible connection to new DEMO VSI] **********************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY [Add team ssh keys] *********************************************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY [Install Tower] *************************************************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   


PLAY [Create IBM Cloud VPC VSI] **************************************************************************************************************************************************************************************************************

TASK [Print some debug information] **********************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": [
        "Requested Topology:",
        "--------------------------------",
        "Topology: legacy-cluster",
        "Inventory Template: inventory_cluster_4.j2",
        "",
        "Control Plane:",
        "--------------------------------",
        "Control Nodes: 0",
        "Hybrid Nodes: 3",
        "Total Control Plane Nodes to deploy: 3",
        "",
        "Execution Plane:",
        "--------------------------------",
        "Hop Nodes: 0",
        "Execution Nodes: 0",
        "Total nodes to deploy for execution plane: 0",
        "Enable execution nodes: False",
        "",
        "",
        "Isolated Nodes:",
        "--------------------------------",
        "Legacy Isolated Nodes requested: 0",
        "Legacy Isolated Nodes to deploy: 0",
        "Enable legacy isolated nodes: False",
        ""
    ]
}

PLAY [Create IBM Cloud VPC VSI for Tower Isolated Nodes] *************************************************************************************************************************************************************************************

PLAY [Create IBM Cloud VPC VSI for Tower Execution Nodes] ************************************************************************************************************************************************************************************

PLAY [Check Ansible connection to new DEMO VSI] **********************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY [Add team ssh keys] *********************************************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY [Install Tower] *************************************************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

```

jainnikhil30 commented on vars.yml:
```diff
@@ -38,7 +111,7 @@ ibmcloud_image_ids:
 
 ibmcloud_vsi_profile: 'bx2-4x16'
-ibmcloud_vsi_count: 2
+ibmcloud_vsi_count: "{{ tower_control_nodes|int + tower_hybrid_nodes|int |int}}"
```
we shouldn't do this. Please note this repo is also used by other users who spin up vm's on ibm cloud without tower being installed. For them tower_control_nodes and tower_hybrid_nodes will not have the relevance.

kdelee replied:
ahh ok

jainnikhil30 commented on vars.yml:
```diff
@@ -2,7 +2,80 @@
+    legacy_isolated_nodes: 0
+    hop_nodes: 0
+    inventory_template: "inventory_cluster_4.j2"
+
```
We should do this some where else may be ? or may be just use the whole dictionary via a different vars file. This repo is used by other teams as well who just create infrastructure (vm's) with tower being installed. They just set install_tower to false like mentioned in the readme.

kdelee replied:
I think ideally we would do 
`automationcontroller <- hop_node[0] <- hop_node[1] .... <- execution_nodes[:]`\

I'll poke at this some more

kdelee replied:
I'm open to moving the tower stuff into a seperate vars file we pass in the jenkins job we use, but I'd move the isolated nodes and execution nodes vars too.

jainnikhil30 replied:
True. I think at this point we should do that. I think that vars.yml is growing into a very Tower/Controller specific file, I am responsible for that. We can keep that file to bare minimum required for spinning up vm's on IBM Cloud without Tower/Controller and move Tower/Controller specific part to different file.

kdelee commented:
@jainnikhil30 I've addressed some of your concerns, not all.

re: these vars being set in this shared vars file. I will contest that there was already alot of tower vars in here

re: use for non-tower installs. I've added two scenarios, one is `plain-nodes`, which is default, and deploys 2 nodes as the old defaults used to. the `custom` scenario sets absolutely no values and you have to override everything. Regardless of what scenario you select, you can still directly set all the vars like `ibmcloud_vsi_count` etc on the command line/in the jenkins job and they will take precedence.

re: hop nodes. I still need to work on that

kdelee commented:
What I've done with the hop nodes in the inventory file now creates a topology like this, where the hop nodes are peered in a chain
    automationcontroller -> hop1 -> hop2 ... -> hopN -> E1,E2,EN

![Screenshot from 2021-11-04 13-01-17](https://user-images.githubusercontent.com/29239279/140390681-cc5b9228-9232-4030-ab94-1081dfbce42c.png)

kdelee commented:
I had a successful-ish run (some tests failed, but this was related to a controller bug where jobs were not finishing)

It got through the whole deploy, deployed the expected topology, and ran the tests.
@jainnikhil30 can you take a look and help me figure out what more we want to do before merging. I already addressed seperating out the vars

kdelee commented:
One more run https://master-jenkins-csb-perf.apps.ocp4.prod.psi.redhat.com/job/sandbox/job/ElijahTowerPerfDevelIBMCloud/38/

jainnikhil30 commented on vars.yml:
```diff
@@ -37,50 +33,15 @@ ibmcloud_image_ids:
-#  - execution
-#  - hybrid
+ibmcloud_vsi_image_id: '{{ ibmcloud_image_ids[ibmcloud_vpc_region][ibmcloud_image] }}'
 
```
Which file are these going into ? Are you moving them into a separate file ?

jainnikhil30 replied:
```    - name: Fetch the tower variables from var file
      include_vars:
        file: install_tower_vars.yml
      when: install_tower|default(false)
      tags: debug```

I am not able to find the  `install_tower_vars.yml`.

kdelee replied:
fixed

jainnikhil30 reviewed (approved):
Thanks for the PR @kdelee

kdelee commented:
OK... finally got past the problem I was having when I had more than 1 hop node...it was a indexing problem in the initial grouping in `create.yml`

NOW...should be able to have arbitrary number of hop nodes/execution nodes and the templating will work fine

Example of control-cluster-2-hop-to-execution-plane
![Screenshot from 2021-12-02 14-47-53](https://user-images.githubusercontent.com/29239279/144492744-fabc048b-2715-4a8c-96d4-ec6392c42d3a.png)

Given that we are past release, I will test a bit more this afternoon, but I feel pretty confident about merging this, so I will do so by end of day given I don't find anything.

If there is any fall out -- I'm 100% ready to hop on it before Dec 20th

kdelee commented:
K, got through a few installs successfully. Going to merge
